File: greg_and_array/ganda.cpp
#include <iostream>
#include <vector>
using namespace std;

void solution(vector<vector<int>>& operations, vector<int>& arr, 
              vector<vector<int>>& queries, int n, int m, int k) {
    // initialize array of size n + 2
    vector<long long> diff_arr(n + 2, 0);
    
    // populate difference array
    for (auto& q : queries) {
        for (int i = q[0]; i <= q[1]; i++) {
            auto& op = operations[i - 1];  // convert to 0-indexed
            diff_arr[op[0]] += op[2];
            diff_arr[op[1] + 1] -= op[2];
        }
    }
    
    // compute prefix sum
    vector<long long> p(n + 1, 0);
    for (int i = 1; i <= n; i++) {
        p[i] = p[i - 1] + diff_arr[i];
    }
    
    // output final array
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] + p[i + 1];
        if (i < arr.size() - 1) cout << " ";
    }
    cout << "\n";
}

int main() {
    int n, m, k;
    cin >> n >> m >> k;
    
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    vector<vector<int>> operations(m, vector<int>(3));
    for (int i = 0; i < m; i++) {
        cin >> operations[i][0] >> operations[i][1] >> operations[i][2];
    }
    
    vector<vector<int>> queries(k, vector<int>(2));
    for (int i = 0; i < k; i++) {
        cin >> queries[i][0] >> queries[i][1];
    }
    
    solution(operations, arr, queries, n, m, k);
    
    return 0;
}

File: greg_and_array/ganda.py
def solution(operations: list[list[int]], arr: list[int], queries: list[list[int]], n: int, m: int, k: int):
    # diff array for the operations
    diff_arr_ops = [0] * (m + 2)
    for q in queries:
        diff_arr_ops[q[0]] += 1
        diff_arr_ops[q[1] + 1] -= 1

    # compute presum for ops
    p = [0] * (m + 1)
    for i in range(1, m+1):
        p[i] = p[i-1] + diff_arr_ops[i]

    # diff array for input 
    diff_arr = [0] * (n + 2)
    for i in range(len(operations)):
        op = operations[i]
        op_num = i + 1  # Convert 0-indexed to 1-indexed operation number
        diff_arr[op[0]] += op[2] * p[op_num]
        diff_arr[op[1] + 1] -= op[2] * p[op_num]

    # compute presum
    p = [0] * (n + 1)
    for i in range(1, n+1):
        p[i] = p[i-1] + diff_arr[i]
    
    # return final array
    for i in range(len(arr)):
        print(arr[i] + p[i+1], end=" ")

def main():
    n, m, k = map(int, input().split())
    arr = list(map(int, input().split()))
    
    operations = []
    for i in range(m):
        operations.append(list(map(int, input().split())))
    
    queries = []
    for i in range(k):
        queries.append(list(map(int, input().split())))
    
    solution(operations, arr, queries, n, m, k)

if __name__ == "__main__":
    main()
